packages/oauth/src/providers/keycloak.ts

```diff
@@ -0,0 +1,253 @@
+import {
+	OAuth2ProviderAuthWithPKCE,
+	createOAuth2AuthorizationUrlWithPKCE,
+	refreshOAuth2Tokens,
+	validateOAuth2AuthorizationCode
+} from "../core/oauth2.js";
+import { ProviderUserAuth } from "../core/provider.js";
+import { decodeIdToken } from "../index.js";
+import { encodeBase64 } from "../utils/encode.js";
+import { handleRequest, authorizationHeader } from "../utils/request.js";
+
+import type { Auth } from "lucia";
+
+type Config = {
+  domain: string;
+  realm: string;
+  clientId: string;
+  clientSecret: string;
+  scope?: string[];
+  redirectUri?: string;
+};
+
+const PROVIDER_ID = "keycloak";
+
+export const keycloak = <_Auth extends Auth = Auth>(
+	auth: _Auth,
+	config: Config
+): KeycloakAuth<_Auth> => {
+	return new KeycloakAuth(auth, config);
+};
+
+export class KeycloakAuth<_Auth extends Auth = Auth> extends OAuth2ProviderAuthWithPKCE<
+	KeycloakUserAuth<_Auth>
+> {
+	private config: Config;
+
+	constructor(auth: _Auth, config: Config) {
+		super(auth);
+
+		this.config = config;
+	}
+
+	public getAuthorizationUrl = async (): Promise<
+		readonly [url: URL, codeVerifier: string, state: string]
+	> => {
+    const scopeConfig = this.config.scope ?? [];
+		return await createOAuth2AuthorizationUrlWithPKCE(
+			`https://${ this.config.domain }/realms/${ this.config.realm }/protocol/openid-connect/auth`,
+			{
+        clientId: this.config.clientId,
+        scope: ["profile", "openid", ...scopeConfig],
+        redirectUri: this.config.redirectUri,
+        codeChallengeMethod: "S256"
+			}
+		);
+	};
+
+	public validateCallback = async (
+		code: string,
+    code_verifier: string
+	): Promise<KeycloakUserAuth<_Auth>> => {
+		const keycloakTokens = await this.validateAuthorizationCode(code, code_verifier);
+		const keycloakUser = await getKeycloakUser(this.config.domain, this.config.realm, keycloakTokens.accessToken);
+		const keycloakRoles = getKeycloakRoles(keycloakTokens.accessToken)
+		return new KeycloakUserAuth(this.auth, keycloakUser, keycloakTokens, keycloakRoles);
+	};
+
+	private validateAuthorizationCode = async (
+		code: string,
+    codeVerifier: string
+	): Promise<KeycloakTokens> => {
+		const rawTokens = await validateOAuth2AuthorizationCode<AccessTokenResponseBody>(
+			code,
+			`https://${ this.config.domain }/realms/${ this.config.realm }/protocol/openid-connect/token`,
+			{
+				clientId: this.config.clientId,
+				redirectUri: this.config.redirectUri,
+				codeVerifier,
+				clientPassword: {
+					authenticateWith: "http_basic_auth",
+					clientSecret: this.config.clientSecret
+				}
+			}
+		);
+
+		return this.claimTokens(rawTokens)
+	};
+
+	public refreshTokens = async (
```

Comment:
I don't hate it, but just for the sake of limiting the scope of the API, I don't want to add it, at least just yet

Reply:
It's a step to support Authorization Providers and Flows which requires this. Everyone could write this if he needs. But that costs a lot of reverse engineering of lucia, while lucia could just provide this method for the sake of Authorization Flows, which require advanced security and ways to provide authorization (roles). Why shouldn't lucia use this potential?

Reply:
`@lucia-auth/oauth` is for handling OAuth. Refresh tokens aren't part of the spec and it increases the maintenance burden (especially since it opens the possibility to do the same for other providers). Like I said, I don't hate it, but I don't want to implement it right now.

Reply:
its inefficient to create methods which do half of the work they should do. This results in rewritting the same code with the extra pepper just to get the function it requires.